### PR TITLE
[AIRFLOW-1502] Implement possibility to search runs

### PR DESCRIPTION
**Pull request by skudriashev**
Implement possibility to search runs via experimental api.

Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following:
    - [AIRFLOW-1502](https://issues.apache.org/jira/browse/AIRFLOW-1502) Implement possibility to search dag runs via experimental api

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

Implement possibility to search dag runs via experimental api.

**Format:**
```
http://localhost:8080/api/experimental/dag_runs/search/<query>
```

**Example:**
```
http://localhost:8080/api/experimental/dag_runs/search/scheduled
```

**Response:**
```
{
  "items": [
    {
      "dag_id": "trigger_1", 
      "dag_run_url": "/admin/airflow/graph?execution_date=2017-08-10+09%3A03%3A00&dag_id=trigger_1", 
      "execution_date": "2017-08-10 09:03", 
      "run_id": "scheduled__2017-08-10T09:03:00", 
      "start_date": "2017-08-10 12:02", 
      "state": "running"
    }, 
    {
      "dag_id": "trigger_1", 
      "dag_run_url": "/admin/airflow/graph?execution_date=2017-08-10+09%3A04%3A00&dag_id=trigger_1", 
      "execution_date": "2017-08-10 09:04", 
      "run_id": "scheduled__2017-08-10T09:04:00", 
      "start_date": "2017-08-10 12:02", 
      "state": "running"
    }
  ]
}
```

### Tests
- [x] My PR adds the following unit tests:
    - tests.models:TestDagRun.test_to_json
    - tests.www.api.experimental.test_endpoints

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: airflow/www/api/experimental/endpoints.py [hunk 1]
```diff
@@ -153,24 +153,21 @@ def task_instance_info(dag_id, execution_date, task_id):
     return jsonify(fields)
 
 
-@api_experimental.route('/latest_runs', methods=['GET'])
+@api_experimental.route('/dag_runs/search/<string:query>', methods=['GET'])
 @requires_authentication
-def latest_dag_runs():
```
Comment (saguziel) [inline airflow/www/api/experimental/endpoints.py]: ? Why did you delete this?

Comment (saguziel) [reply airflow/www/api/experimental/endpoints.py]: / change it

File: airflow/www/api/experimental/endpoints.py [hunk 2]
```diff
@@ -153,24 +153,21 @@ def task_instance_info(dag_id, execution_date, task_id):
     return jsonify(fields)
 
 
-@api_experimental.route('/latest_runs', methods=['GET'])
+@api_experimental.route('/dag_runs/search/<string:query>', methods=['GET'])
```
Comment (saguziel) [inline airflow/www/api/experimental/endpoints.py]: in hindsight, we should probably have it as `dagruns` and not `dag_runs`

File: airflow/www/templates/airflow/dags.html [hunk 3]
```diff
@@ -226,7 +226,7 @@
           }
         });
       });
-      $.getJSON("{{ url_for('api_experimental.latest_dag_runs') }}", function(data) {
+      $.getJSON("{{ url_for('api_experimental.dag_runs_latest') }}", function(data) {
```
Comment (saguziel) [inline airflow/www/templates/airflow/dags.html]: Does not have a dag_run_url field, and execution_date may not be formatted correctly, along with start_date processing not being there

File: airflow/www/api/experimental/endpoints.py (hunk 2 above)
Comment (skudriashev) [reply airflow/www/api/experimental/endpoints.py]: @saguziel why?

File: airflow/www/api/experimental/endpoints.py (hunk 1 above)
Comment (skudriashev) [reply airflow/www/api/experimental/endpoints.py]: I renamed it to be consistent with other function.

File: airflow/www/templates/airflow/dags.html (hunk 3 above)
Comment (skudriashev) [reply airflow/www/templates/airflow/dags.html]: sorry? I just renamed endpoint, nothing else.

File: airflow/www/api/experimental/endpoints.py (hunk 2 above)
Comment (saguziel) [reply airflow/www/api/experimental/endpoints.py]: Because the class is called DagRun and python is usually configuration over convention

File: airflow/www/api/experimental/endpoints.py (hunk 1 above)
Comment (saguziel) [reply airflow/www/api/experimental/endpoints.py]: You changed the behavior

File: airflow/www/templates/airflow/dags.html (hunk 3 above)
Comment (saguziel) [reply airflow/www/templates/airflow/dags.html]: You changed the behavior

Comment (skudriashev) [reply airflow/www/templates/airflow/dags.html]: I changed to be consistent.

File: airflow/www/api/experimental/endpoints.py (hunk 1 above)
Comment (skudriashev) [reply airflow/www/api/experimental/endpoints.py]: Yes I changed it.

File: airflow/www/api/experimental/endpoints.py (hunk 2 above)
Comment (skudriashev) [reply airflow/www/api/experimental/endpoints.py]: Do you want me to change to `dagruns`?

**Comment by skudriashev** [comment]
@saguziel, so all good? :)

**Comment by skudriashev** [comment]
@r39132 @criccomini @bolkedebruin will you have time to review this? Please.

**Comment by bolkedebruin** [comment]
Please address @saguziel points or clarify (explain elaborately) why you want some changes. If you are changing endpoints for example then this has an effect on running systems. It then needs to be properly documented and marked for a major release. Only then you can get a +1

**Comment by skudriashev** [comment]
@bolkedebruin ok, I see your point. I can keep existing one as-is. Will update this PR.

**Comment by skudriashev** [comment]
@bolkedebruin updated, please review. Thanks)

**Comment by skudriashev** [comment]
@bolkedebruin, please review. Thanks!

File: airflow/models.py [hunk 4]
```diff
@@ -4363,6 +4364,21 @@ def __repr__(self):
             run_id=self.run_id,
             external_trigger=self.external_trigger)
 
+    def to_json(self):
```
Comment (bolkedebruin) [inline airflow/models.py:4367]: Please remove the dependency on flask in models.py. I don't think it makes sense to connect the model to the view this way. This is basically a hard coded url.

I'm not convinced to_json belongs here. It is also not a full serialisation of the object.

**Comment by bolkedebruin** [comment]
This PR should follow the generic architecture for APIs:

* Please make sure business logic is in airflow.api.common.experimental
* Expose API definitions in airflow.api.client.json/local 

Do not add business logic to models (search should probably be in common.experimental.search_dagrun.py)

**Comment by skudriashev** [comment]
@bolkedebruin thanks! Will fix.
